File: market-mapper/src/main/java/by/market/mapper/ProductMapper.java
package by.market.mapper;

import by.market.domain.Product;
import by.market.dto.product.ProductDTO;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(config = MapperConfiguration.class)
public interface ProductMapper extends MapstructMapper<ProductDTO, Product> {

    @Override
    @Mapping(source = "category.id", target = "category")
    ProductDTO toMap(Product product);

    @Override
    @Mapping(source = "category", target = "category.id")
    Product fromMap(ProductDTO productDTO);

}

File: market-mapper/src/main/kotlin/by/market/mapper/product/ProductMapper.kt
package by.market.mapper.product

import by.market.domain.Product
import by.market.dto.product.ProductDTO
import by.market.mapper.IMapstructMapper
import by.market.mapper.MapperConfig
import org.mapstruct.Mapper
import org.mapstruct.Mapping

@Mapper(config = MapperConfig::class)
interface ProductMapper : IMapstructMapper<ProductDTO, Product> {

    @Mapping(source = "category.id", target = "category")
    override fun to(e: Product): ProductDTO

    @Mapping(source = "category", target = "category.id")
    override fun from(d: ProductDTO): Product

}
